File: 01. Arrays/09. Majority Element II/main.java
import java.util.*;

class main {
    // Brute Force Approach - O(n²) Time, O(1) Space
    public static List<Integer> majorityElementBruteForce(int[] nums) {
        List<Integer> result = new ArrayList<>();
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = 0; j < n; j++) {
                if (nums[j] == nums[i]) count++;
            }
            if (count > n / 3 && !result.contains(nums[i])) result.add(nums[i]);
        }
        return result;
    }

    // Sorting Approach - O(n log n) Time, O(1) Space
    public static List<Integer> majorityElementSorting(int[] nums) {
        Arrays.sort(nums);
        List<Integer> result = new ArrayList<>();
        int n = nums.length;
        for (int i = 0; i < n;) {
            int count = 0, num = nums[i];
            while (i < n && nums[i] == num) { count++; i++; }
            if (count > n / 3) result.add(num);
        }
        return result;
    }

    // HashMap Approach - O(n) Time, O(n) Space
    public static List<Integer> majorityElementHashMap(int[] nums) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int num : nums) freq.put(num, freq.getOrDefault(num, 0) + 1);
        List<Integer> result = new ArrayList<>();
        for (int key : freq.keySet()) {
            if (freq.get(key) > nums.length / 3) result.add(key);
        }
        return result;
    }

    // Boyer-Moore Voting Algorithm - O(n) Time, O(1) Space
    public static List<Integer> majorityElementBoyerMoore(int[] nums) {
        int candidate1 = 0, candidate2 = 0, count1 = 0, count2 = 0;
        for (int num : nums) {
            if (num == candidate1) count1++;
            else if (num == candidate2) count2++;
            else if (count1 == 0) { candidate1 = num; count1 = 1; }
            else if (count2 == 0) { candidate2 = num; count2 = 1; }
            else { count1--; count2--; }
        }
        count1 = count2 = 0;
        for (int num : nums) {
            if (num == candidate1) count1++;
            else if (num == candidate2) count2++;
        }
        List<Integer> result = new ArrayList<>();
        int n = nums.length;
        if (count1 > n / 3) result.add(candidate1);
        if (count2 > n / 3) result.add(candidate2);
        return result;
    }

    // Main Function
    public static void main(String[] args) {
        int[] nums = {3, 2, 3, 1, 1, 1, 2, 2};
        System.out.println("Brute Force: " + majorityElementBruteForce(nums));
        System.out.println("Sorting: " + majorityElementSorting(nums));
        System.out.println("HashMap: " + majorityElementHashMap(nums));
        System.out.println("Boyer-Moore: " + majorityElementBoyerMoore(nums));
    }
}

File: 01. Arrays/09. Majority Element II/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Brute Force
vector<int> majorityElementBruteForce(vector<int>& nums) {
    int n = nums.size();
    vector<int> result;
    for (int i = 0; i < n; i++) {
        int count = 0;
        for (int j = 0; j < n; j++) {
            if (nums[j] == nums[i]) count++;
        }
        if (count > n / 3 && find(result.begin(), result.end(), nums[i]) == result.end()) {
            result.push_back(nums[i]);
        }
    }
    return result;
}

// Sorting Approach
vector<int> majorityElementSorting(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    int n = nums.size();
    vector<int> result;
    for (int i = 0; i < n;) {
        int count = upper_bound(nums.begin(), nums.end(), nums[i]) - lower_bound(nums.begin(), nums.end(), nums[i]);
        if (count > n / 3) result.push_back(nums[i]);
        i += count;
    }
    return result;
}

// HashMap Approach
vector<int> majorityElementHashMap(vector<int>& nums) {
    unordered_map<int, int> freq;
    for (int num : nums) freq[num]++;
    vector<int> result;
    for (auto [key, value] : freq) {
        if (value > nums.size() / 3) result.push_back(key);
    }
    return result;
}

// Boyer-Moore Voting Algorithm
vector<int> majorityElementBoyerMoore(vector<int>& nums) {
    int candidate1 = 0, candidate2 = 0, count1 = 0, count2 = 0;
    for (int num : nums) {
        if (num == candidate1) count1++;
        else if (num == candidate2) count2++;
        else if (count1 == 0) candidate1 = num, count1 = 1;
        else if (count2 == 0) candidate2 = num, count2 = 1;
        else count1--, count2--;
    }

    count1 = count2 = 0;
    for (int num : nums) {
        if (num == candidate1) count1++;
        else if (num == candidate2) count2++;
    }

    vector<int> result;
    int n = nums.size();
    if (count1 > n / 3) result.push_back(candidate1);
    if (count2 > n / 3) result.push_back(candidate2);
    
    return result;
}

// Main Function
int main() {
    vector<int> nums = {3, 2, 3, 1, 1, 1, 2, 2};
    cout << "Brute Force: "; for (int num : majorityElementBruteForce(nums)) cout << num << " ";
    cout << "\nSorting: "; for (int num : majorityElementSorting(nums)) cout << num << " ";
    cout << "\nHashMap: "; for (int num : majorityElementHashMap(nums)) cout << num << " ";
    cout << "\nBoyer-Moore: "; for (int num : majorityElementBoyerMoore(nums)) cout << num << " ";
    return 0;
}
